### ISSUE TITLE
Segmentation Fault with Pandas

### ISSUE THREAD

**Issue by EthanRosenthal** [maciejkula/spotlight#42]
I ran into a very odd segmentation fault error. This could very well be a PyTorch bug, but I thought I'd bring it up here, first. I've produced a minimal example at the bottom of this issue.

So far, I know that the fault happens at the `loss.backward()` call in `model.fit()`. The fault only seems to happen under the combination of two conditions (that I can find, so far):
1. When `sparse=True`. 
2. Pandas is imported at the top of the file

(BTW, I pass in an SGD optimizer because that seems to be the only one that works right now with sparse embeddings)

I'm using pandas version 0.20.3 from conda, the latest spotlight from master, and PyTorch 0.2.0 from conda. I'd love to know if others can reproduce this.

As I said, this could very well be a PyTorch bug, but, if others run into this, it'll be helpful to have this issue as a reference.

```python
import pandas as pd
import numpy as np
import torch

from spotlight.interactions import Interactions
from spotlight.factorization.implicit import ImplicitFactorizationModel

user_ids = [2471, 5808, 3281, 4086, 6293, 8970, 11828, 3281]
item_ids = [1583, 57, 6963, 867, 8099, 10991, 24, 800]
num_users = 15274
num_items = 25655

train = Interactions(np.array(user_ids, dtype=np.int64),
                     np.array(item_ids, dtype=np.int64),
                     num_users=num_users,
                     num_items=num_items)

def optimizer_func(params, lr=0.01):
    return torch.optim.SGD(params, lr=lr)
  
RANDOM_STATE = np.random.RandomState(42)
model = ImplicitFactorizationModel(loss='bpr',
                                   embedding_dim=32,
                                   batch_size=4,
                                   n_iter=1,
                                   use_cuda=False,
                                   optimizer_func=optimizer_func,
                                   sparse=True,
                                   random_state=RANDOM_STATE)
# Fault
model.fit(train, verbose=True)
```

**Comment by maciejkula** [comment]
I can absolutely reproduce. This is what gdb says:
```
42      Python/errors.c: No such file or directory.
(gdb) backtrace
#0  PyErr_Restore (type=0x0, value=<optimised out>, traceback=0x0) at Python/errors.c:42
#1  0x00007ffff7985b9d in PyDict_GetItem (op=0x7fffe696a360, key=0x7fffc3540b70)
    at Objects/dictobject.c:1424
#2  0x00007ffff5cd32ba in _buffer_clear_info (arr=<optimised out>)
    at numpy/core/src/multiarray/buffer.c:610
#3  _array_dealloc_buffer_info (self=<optimised out>) at numpy/core/src/multiarray/buffer.c:776
#4  0x00007ffff5c9713e in array_dealloc (self=<optimised out>)
    at numpy/core/src/multiarray/arrayobject.c:381
#5  0x00007fffe4214acd in THPPointer<_object>::operator= (new_ptr=0x0, this=0x1793ff0)
   from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so
#6  NumpyArrayAllocator::free (this=0x1793ff0, ptr=<optimised out>) at torch/csrc/allocators.cpp:71
#7  0x00007fffddb7f74e in THLongStorage_free ()
   from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/lib/libTH.so.1
#8  0x00007fffddba0374 in THLongTensor_free ()
   from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/lib/libTH.so.1
#9  0x00007fffdd92d3f1 in THSFloatTensor_free ()
   from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/lib/libTHS.so.1
#10 0x00007fffdd5aff10 in thpp::THSTensor<float>::~THSTensor() ()
   from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/lib/libTHPP.so.1
#11 0x00007fffdd5aff29 in thpp::THSTensor<float>::~THSTensor() ()
   from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/lib/libTHPP.so.1
#12 0x00007fffe4225151 in std::default_delete<thpp::Tensor>::operator() (this=<optimised out>, 
    __ptr=<optimised out>)
   from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so
#13 std::unique_ptr<thpp::Tensor, std::default_delete<thpp::Tensor> >::~unique_ptr (
    this=0x7fffbc001a78, __in_chrg=<optimised out>) at torch/csrc/autograd/variable.cpp:184
#14 torch::autograd::Variable::~Variable (this=0x7fffbc001a68, __in_chrg=<optimised out>)
    at torch/csrc/autograd/variable.cpp:17
#15 __gnu_cxx::new_allocator<torch::autograd::Variable>::destroy<torch::autograd::Variable> (this=<optimised out>, __p=0x7fffbc001a68) at torch/csrc/autograd/variable.cpp:124
#16 std::allocator_traits<std::allocator<torch::autograd::Variable> >::_S_destroy<torch::autograd::Variable> (__a=..., __p=0x7fffbc001a68) at torch/csrc/autograd/variable.cpp:281
#17 std::allocator_traits<std::allocator<torch::autograd::Variable> >::destroy<torch::autograd::Variable> (__a=..., __p=0x7fffbc001a68) at torch/csrc/autograd/variable.cpp:405
#18 std::_Sp_counted_ptr_inplace<torch::autograd::Variable, std::allocator<torch::autograd::Variable>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (this=<optimised out>) at torch/csrc/autograd/variable.cpp:407
#19 0x00007fffe421cbe9 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x7fffbc001a50) from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so
#20 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x7fffbc002f78, __in_chrg=<optimised out>) at torch/csrc/autograd/engine.cpp:546
#21 std::__shared_ptr<torch::autograd::Variable, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fffbc002f70, __in_chrg=<optimised out>) at torch/csrc/autograd/engine.cpp:781
#22 std::shared_ptr<torch::autograd::Variable>::~shared_ptr (this=0x7fffbc002f70, __in_chrg=<optimised out>) at torch/csrc/autograd/engine.cpp:93
#23 std::_Destroy<std::shared_ptr<torch::autograd::Variable> > (__pointer=0x7fffbc002f70) at torch/csrc/autograd/engine.cpp:93
#24 std::_Destroy_aux<false>::__destroy<std::shared_ptr<torch::autograd::Variable>*> (__last=<optimised out>, __first=0x7fffbc002f70) at torch/csrc/autograd/engine.cpp:103
#25 std::_Destroy<std::shared_ptr<torch::autograd::Variable>*> (__last=<optimised out>, __first=<optimised out>) at torch/csrc/autograd/engine.cpp:126
#26 std::_Destroy<std::shared_ptr<torch::autograd::Variable>*, std::shared_ptr<torch::autograd::Variable> > (__last=0x7fffbc002f80, __first=0x7fffbc002f60) at torch/csrc/autograd/engine.cpp:151
#27 std::vector<std::shared_ptr<torch::autograd::Variable>, std::allocator<std::shared_ptr<torch::autograd::Variable> > >::~vector (this=0x7fffc1773bd0, __in_chrg=<optimised out>)
    at torch/csrc/autograd/engine.cpp:415
#28 torch::autograd::Engine::evaluate_function (this=this@entry=0x7fffe523d140 <engine>, task=...) at torch/csrc/autograd/engine.cpp:232
#29 0x00007fffe421d9e9 in torch::autograd::Engine::thread_main (this=this@entry=0x7fffe523d140 <engine>, queue=..., device=device@entry=-1)
   from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so
#30 0x00007fffe42399ea in PythonEngine::thread_main (this=0x7fffe523d140 <engine>, queue=..., device=-1) from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so
#31 0x00007fffe42221ce in std::_Mem_fn<void (torch::autograd::Engine::*)(std::shared_ptr<torch::autograd::ReadyQueue>, int)>::operator()<std::shared_ptr<torch::autograd::ReadyQueue>, int, void> (
    __object=<optimised out>, this=<optimised out>) from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so
#32 std::_Bind_simple<std::_Mem_fn<void (torch::autograd::Engine::*)(std::shared_ptr<torch::autograd::ReadyQueue>, int)>(torch::autograd::Engine*, std::shared_ptr<torch::autograd::ReadyQueue>, int)>::_M_invoke<0ul, 1ul, 2ul> (this=<optimised out>) at torch/csrc/autograd/e
ngine.cpp:1732
#33 std::_Bind_simple<std::_Mem_fn<void (torch::autograd::Engine::*)(std::shared_ptr<torch::autograd::ReadyQueue>, int)> (torch::autograd::Engine*, std::shared_ptr<torch::autograd::ReadyQueue>, int)>::operator()() (this=<optimised out>) at torch/csrc/autograd/engine.cpp:1
720
#34 std::thread::_Impl<std::_Bind_simple<std::_Mem_fn<void (torch::autograd::Engine::*)(std::shared_ptr<torch::autograd::ReadyQueue>, int)> (torch::autograd::Engine*, std::shared_ptr<torch::autograd::ReadyQueue>, int)> >::_M_run() (this=<optimised out>) at torch/csrc/auto
grad/engine.cpp:115
#35 0x00007fffe4d65840 in execute_native_thread_routine () from /home/maciej/miniconda3/lib/python3.6/site-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so
#36 0x00007ffff76bc6ba in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#37 0x00007ffff6ada3dd in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

**Comment by maciejkula** [comment]
It looks like it's `autograd`, it triggers on the `loss.backward()` call.

**Comment by maciejkula** [comment]
Right, it feels like calling `PyThreadState_GET()` returns a null pointer here: https://github.com/python/cpython/blob/5fd33b5926eb8c9352bf5718369b4a8d72c4bb44/Python/errors.c#L28

Examining in GDB:
```
(gdb) down
#0  PyErr_Restore (type=0x0, value=<optimised out>, traceback=0x0) at Python/errors.c:42
42      Python/errors.c: No such file or directory.
(gdb) info locals
tstate = 0x0
oldtype = <optimised out>
oldvalue = <optimised out>
oldtraceback = <optimised out>
(gdb) tstate
Undefined command: "tstate".  Try "help".
(gdb) p tstate
```

**Comment by EthanRosenthal** [comment]
Thanks for diving in! I'll avoid importing Pandas for now :slightly_smiling_face: 

And I think your suspicion is right. [Looks](https://github.com/pandas-dev/pandas/issues/9394) like there's lots of "thready" things that happen when Pandas gets imported.

**Comment by apaszke** [comment]
A fix will be soon in master - https://github.com/pytorch/pytorch/pull/2581 Sorry for keeping you waiting!

**Comment by maciejkula** [comment]
@apaszke Awesome. Thanks for the amazing work on PyTorch!
